### ISSUE TITLE
spectral profile from a point region is not stoppable 

### ISSUE THREAD

**Issue by kswang1029** [CARTAvis/carta-backend#298]
with dev as of July 31, 2019

the backend keeps sending updates of a point region spectral profile even if the user switches to another region.

UPDATE: same for cursor region.

**Comment by kswang1029** [comment]
shall we consider this as critical (from the usability point of view if the image is big)? This means users always have to wait the cursor profile finished then start to see region profile updating. When this happens, the spectral profiler widget appears _blank_ until the region profile starts to update.

**Comment by markccchiang** [comment]
In PR #288, I have considered the current region id which can be used to cancel the cursor or the other region id spectral profile calculation. However, if we open additional zprofile widgets for the other regions, then only the widgets with specific region id can be updated. Is it acceptable for you?

**Comment by kswang1029** [comment]
@markccchiang if you are referring to "origin/mark/refactor_profile_cancellation 296dd45 5027 2019-08-05 15:58:50 CST", cursor profile is still non-stoppable.

**Comment by markccchiang** [comment]
This function was removed according to Pam's suggestion. Maybe you can test 2a9413619e2d5a7ccafe9f938e392d5f4d346ecf.

**Comment by kswang1029** [comment]
@markccchiang sorry with the mix n' match page, 2a94136 is not available anymore. Is it merged?

**Comment by markccchiang** [comment]
It is not merged yet. Should I move the current region id as an interruption condition back?

**Comment by kswang1029** [comment]
The desired feature is the following. If we have three regions: cursor, point, and rectangle, and users request cursor spectral profile, then if users switch to either point or rectangle region while the cursor profile is updating, the cursor profile should stop updating immediately and start the new profile calculations for the new region request.

**Comment by pford** [comment]
If the user changes the profile widget cursor->region X, the frontend sends a new cursor spectral requirement message with no profiles, which triggers cancelling the cursor profile.  Or perhaps you are saying to prefer the other region over cursor when both are required, but that will require some bookkeeping to resume the cursor profile once the other region profile is done.

**Comment by veggiesaurus** [comment]
I think what @kswang1029 is saying is that if the cursor profile is no longer required, then it should be cancelled. If it is still required by another widget, it shouldn't be cancelled. I believe this is the intended behaviour of the current implementation. 

If a region and a cursor profile are required, then they're calculated one by one, due to the casacore mutex. One way to calculate them "simultaneously" is to lock and unlock the mutex during each iteration of the profile calculation, but that's likely to cause both profiles to take longer, as it adds mutex overhead, and thrashing I/O

Currently, if there are too many blocked tasks, then tbb has no more threads to allocate tasks, and any new SET_CURSOR or SET_SPECTRAL_REQUIREMENTS messages just get queued, rather than being processed and interrupting the currently calculating spectral profile.

**Comment by markccchiang** [comment]
The backend needs to know the zprofile widgets with respect to their region id, stats type and stoke index shown on the frontend. Then use this information to decide which zprofile calculations need to be canceled.

**Comment by pford** [comment]
Perhaps RegionProfiler::DiffSpectralRequirements can be extended to determine which profiles to cancel in addition to which profiles to add (accessed with RegionProfiler::GetSpectralStatsToLoad)?

**Comment by pford** [comment]
Merged into dev

**Comment by kswang1029** [comment]
Yes, looking good.
